Disable "Take Picture" button on zooming in

egovova11 opened the pull request:
- When user zooms in, "Take picture" button is disabled, and "You cannot take a picture while zooming" toast is displayed

eduardb commented:
Hi @egovova11, I see that at some point you committed a UI test for this functionality, but then you reverted the commit. Did you encounter some difficulty with the test, or why didn't you keep it? Let me know if there's anything you need help with.

eduardb commented on app/src/main/java/net/aiscope/gdd_app/ui/capture/CaptureImageActivity.kt:
```diff
@@ -54,6 +55,23 @@ class CaptureImageActivity : AppCompatActivity(), CaptureImageView, CaptureFlow
                 cameraErrorCallback = { presenter.onCaptureError(it) }
             )
             zoomController = ZoomController(fotoapparat, cameraZoomLevel, cameraView)
+            toast  = Toast.makeText(
```
This `toast` is going to hold a reference to this Activity, so it means that it needs to be unreferenced in `onDestroy()` to avoid memory leaks.

egovova11 replied:
I'm not sure how to do this properly, but take a look at my attempt

eduardb commented on app/src/main/java/net/aiscope/gdd_app/ui/capture/CaptureImageActivity.kt:
```diff
@@ -37,6 +37,7 @@ class CaptureImageActivity : AppCompatActivity(), CaptureImageView, CaptureFlow
     private lateinit var zoomController: ZoomController
 
     private lateinit var binding: ActivityCaptureImageBinding
+    private lateinit var toast: Toast;
```
No semicolons in Kotlin, please 😛 Unless it's for separating multiple statements on the same line.
```suggestion
    private lateinit var toast: Toast
```

eduardb commented on app/src/main/java/net/aiscope/gdd_app/ui/capture/ZoomController.kt:
```diff
@@ -14,6 +14,8 @@ class ZoomController(
     private val cameraView: View
 ) {
 
+    var onZoomChangedListener : OnZoomChangedListener? = null
```
Why not move this to the constructor? I don't see it ever reassigned, so it could be a `val` too.

egovova11 replied:
done

eduardb commented on app/src/main/java/net/aiscope/gdd_app/ui/capture/ZoomController.kt:
```diff
@@ -98,4 +104,15 @@ class ZoomController(
     }
+
+    private fun onZoomChanged(zoom: Float) {
+        // it is assumed that when the value provided here is equal to 1.0
+        // then the image is not zoomed
+        // and also that zoom cannot be less than 1.0
```
Excellent comment 👍

eduardb commented on app/src/main/java/net/aiscope/gdd_app/ui/capture/ZoomController.kt:
```diff
@@ -98,4 +104,15 @@ class ZoomController(
+        // it is assumed that when the value provided here is equal to 1.0
+        // then the image is not zoomed
+        // and also that zoom cannot be less than 1.0
+        onZoomChangedListener?.onZoomChanged(zoom > 1.0f);
```
May I suggest using a named argument here for more clarity? Also no semicolon, please :P
```suggestion
        onZoomChangedListener?.onZoomChanged(isZoomed = zoom > 1.0f)
```

eduardb commented on app/src/main/java/net/aiscope/gdd_app/ui/capture/ZoomController.kt:
```diff
@@ -98,4 +104,15 @@ class ZoomController(
+    }
+
+    fun interface OnZoomChangedListener {
+        fun onZoomChanged(isZoomed: Boolean) : Unit;
```
Semicolon! 
```suggestion
        fun onZoomChanged(isZoomed: Boolean) : Unit
```

eduardb commented on app/src/main/java/net/aiscope/gdd_app/ui/capture/CaptureImageActivity.kt:
```diff
@@ -54,6 +55,23 @@ class CaptureImageActivity : AppCompatActivity(), CaptureImageView, CaptureFlow
+            toast  = Toast.makeText(
+                this@CaptureImageActivity,
+                R.string.you_cannot_take_a_picture_while_zooming,
+                Toast.LENGTH_SHORT);
+
+            zoomController.onZoomChangedListener = ZoomController.OnZoomChangedListener {
+                if (it)
+                {
+                    captureImageButton.isEnabled = false;
+                    toast.show()
+                }
+                else
+                {
+                    captureImageButton.isEnabled = true;
```
I think that using an explicit parameter would make this more readable, do you agree? Also, please stick to the project conventions of where to place the curly braces, and no semicolons, for consistency reasons.

```suggestion
                Toast.LENGTH_SHORT)

            zoomController.onZoomChangedListener = ZoomController.OnZoomChangedListener { isZoomed ->
                if (isZoomed) {
                    captureImageButton.isEnabled = false
                    toast.show()
                } else {
                    captureImageButton.isEnabled = true
```

eduardb reviewed (changes requested):
This looks good overall, but I have some suggestions, nothing major.

egovova11 commented:
> Hi @egovova11, I see that at some point you committed a UI test for this functionality, but then you reverted the commit. Did you encounter some difficulty with the test, or why didn't you keep it? Let me know if there's anything you need help with.

@eduardb Sorry, I missed your comment initially. The problem was that I was misled by the task description and implemented the logic for a different activity, not the "Take photo", Then I reverted the change and implemented it right. I'm not quite sure how to do the android tests properly for photo activity.
